File: lambdas/default/repositories/game_repo.py
# repositories/game_repo.py

import time

from boto3.dynamodb.conditions import Key, Attr
from botocore.exceptions import ClientError

from mappers.game_mapper import GameMapper

class GameRepository:
    def __init__(self, table):
        self.table = table
    
    def get(self, room_id):
        resp = self.table.get_item(Key={"roomId": room_id})
        item = resp.get("Item")
        if item:
            item = GameMapper.from_dynamo(item)
            return item
        
        return None

    def create(self, item):
        item = GameMapper.to_dynamo(item)
        print(item)
        self.table.put_item(Item=item,ConditionExpression="attribute_not_exists(roomId)")
    
    def update(self, item):
        item = GameMapper.to_dynamo(item)

        update_expr_parts = []
        expr_attr_values = {}
        expr_attr_names = {}
        
        if item.get("status") is not None:
            update_expr_parts.append("#s = :status")
            expr_attr_names["#s"] = "status"
            expr_attr_values[":status"] = item["status"]
        
        if item.get("createdAt") is not None:
            update_expr_parts.append("#c = :created")
            expr_attr_names["#c"] = "createdAt"
            expr_attr_values[":created"] = item["createdAt"]

        if item.get("ttl") is not None:
            update_expr_parts.append("#ttl = :ttl")
            expr_attr_names["#ttl"] = "ttl"
            expr_attr_values[":ttl"] = item["ttl"]

        if item.get("board") is not None:
            update_expr_parts.append("#b = :board")
            expr_attr_names["#b"] = "board"
            expr_attr_values[":board"] = item["board"]

        if item.get("turn") is not None:
            update_expr_parts.append("#t = :turn")
            expr_attr_names["#t"] = "turn"
            expr_attr_values[":turn"] = item["turn"]

        # 🔒 No actualizar si no hay nada
        if not update_expr_parts:
            return

        update_expression = "SET " + ", ".join(update_expr_parts)

        self.table.update_item(
            Key={"roomId": item.get("roomId")},
            UpdateExpression=update_expression,
            ExpressionAttributeNames=expr_attr_names if expr_attr_names else None,
            ExpressionAttributeValues=expr_attr_values
        )
    
    def delete(self, room_id):
        now = int(time.time())
        self.table.update_item(
            Key={"playerId": room_id},
            UpdateExpression="SET #s = :d, ttl = :ttl",
            ConditionExpression="#s <> :d",
            ExpressionAttributeNames={"#s": "status"},
            ExpressionAttributeValues={":d": "GAME OVER", ":ttl": now + 10}
        )

        

File: apiRest/packages/domain/src/repositories/gameRepo.js
"use strict";
// repositories/gameRepo.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.GameRepository = void 0;
const lib_dynamodb_1 = require("@aws-sdk/lib-dynamodb");
const client_dynamodb_1 = require("@aws-sdk/client-dynamodb");
const gameMapper_1 = require("../mappers/gameMapper");
class GameRepository {
    constructor(tableName) {
        this.tableName = tableName;
        this.db = lib_dynamodb_1.DynamoDBDocumentClient.from(new client_dynamodb_1.DynamoDBClient({}));
    }
    async get(roomId) {
        const resp = await this.db.send(new lib_dynamodb_1.GetCommand({
            TableName: this.tableName,
            Key: { roomId },
        }));
        if (!resp.Item)
            return null;
        return gameMapper_1.GameMapper.fromDynamo(resp.Item);
    }
    async create(item) {
        const dynamoItem = gameMapper_1.GameMapper.toDynamo(item);
        console.log(dynamoItem);
        await this.db.send(new lib_dynamodb_1.PutCommand({
            TableName: this.tableName,
            Item: dynamoItem,
            ConditionExpression: "attribute_not_exists(roomId)",
        }));
    }
    async update(item) {
        const dynamoItem = gameMapper_1.GameMapper.toDynamo(item);
        const updateParts = [];
        const exprAttrValues = {};
        const exprAttrNames = {};
        if (dynamoItem.status !== undefined) {
            updateParts.push("#s = :status");
            exprAttrNames["#s"] = "status";
            exprAttrValues[":status"] = dynamoItem.status;
        }
        if (dynamoItem.createdAt !== undefined) {
            updateParts.push("#c = :created");
            exprAttrNames["#c"] = "createdAt";
            exprAttrValues[":created"] = dynamoItem.createdAt;
        }
        if (dynamoItem.ttl !== undefined) {
            updateParts.push("#ttl = :ttl");
            exprAttrNames["#ttl"] = "ttl";
            exprAttrValues[":ttl"] = dynamoItem.ttl;
        }
        if (dynamoItem.board !== undefined) {
            updateParts.push("#b = :board");
            exprAttrNames["#b"] = "board";
            exprAttrValues[":board"] = dynamoItem.board;
        }
        if (dynamoItem.turn !== undefined) {
            updateParts.push("#t = :turn");
            exprAttrNames["#t"] = "turn";
            exprAttrValues[":turn"] = dynamoItem.turn;
        }
        // 🔒 no actualizar si no hay nada
        if (updateParts.length === 0)
            return;
        const updateExpression = `SET ${updateParts.join(", ")}`;
        await this.db.send(new lib_dynamodb_1.UpdateCommand({
            TableName: this.tableName,
            Key: { roomId: dynamoItem.roomId },
            UpdateExpression: updateExpression,
            ExpressionAttributeNames: exprAttrNames,
            ExpressionAttributeValues: exprAttrValues,
        }));
    }
    async delete(roomId) {
        const now = Math.floor(Date.now() / 1000);
        await this.db.send(new lib_dynamodb_1.UpdateCommand({
            TableName: this.tableName,
            Key: { roomId },
            UpdateExpression: "SET #s = :d, ttl = :ttl",
            ConditionExpression: "#s <> :d",
            ExpressionAttributeNames: {
                "#s": "status",
            },
            ExpressionAttributeValues: {
                ":d": "GAME OVER",
                ":ttl": now + 10,
            },
        }));
    }
}
exports.GameRepository = GameRepository;
